client.go

```diff
@@ -66,85 +66,130 @@ func (c *Client) Close() error {
-		}
 		if !bytes.Equal(eth.Destination, c.ifi.HardwareAddr) {
 			continue
 		}
-		if eth.EtherType != ethernet.EtherTypeARP {
+		if !c.IsReplyFrom(ip, arp) {
 			continue
 		}
 
-		// Unmarshal ARP packet and check:
-		//   - Packet is a reply, not a request
-		//   - Packet is for our IP address
-		//   - Packet is for our hardware address
-		//   - Packet is a reply to our query, not another query
-		if err := arp.UnmarshalBinary(eth.Payload); err != nil {
-			return nil, err
-		}
-		if arp.Operation != OperationReply {
-			continue
-		}
-		if !arp.TargetIP.Equal(c.ip) {
-			continue
+		return arp.SenderHardwareAddr, nil
+	}
+}
+
+// IsReply reports whether the packet p is a response to one of our
+// queries. A packet is a response if it is addressed to our IP- and
+// hardware address.
+func (c *Client) IsReply(p *Packet) bool {
+	return p.Operation == OperationReply &&
+		p.TargetIP.Equal(c.ip) &&
+		bytes.Equal(p.TargetHardwareAddr, c.ifi.HardwareAddr)
+}
+
+// IsReplyFrom reports whether the packet p is a response to our query
+// for ip. It works like IsReply, but additionally checks if the
+// response is for a specific IPv4 address ip.
+func (c *Client) IsReplyFrom(ip net.IP, p *Packet) bool {
+	return c.IsReply(p) && ip.Equal(p.SenderIP)
+}
+
+// Read reads a single ARP packet and returns it, together with its
+// ethernet frame.
+func (c *Client) Read() (*Packet, *ethernet.Frame, error) {
+	buf := make([]byte, 128)
+	for {
+		n, addr, err := c.p.ReadFrom(buf)
+		if err != nil {
+			return nil, nil, err
 		}
-		if !bytes.Equal(arp.TargetHardwareAddr, c.ifi.HardwareAddr) {
-			continue
+
+		p, eth, err := parsePacket(buf[:n])
+		if err != nil {
+			if err == errInvalidARPPacket {
+				continue
+			}
+			return nil, nil, err
 		}
-		if !ip.Equal(arp.SenderIP) {
-			continue
+		if addr, ok := addr.(*raw.Addr); ok {
+			p.RemoteAddr = addr.HardwareAddr
 		}
+		return p, eth, nil
+	}
+}
 
-		return arp.SenderHardwareAddr, nil
+// WriteTo writes a single ARP packet to addr. Note that addr should,
+// but doesn't have to, match the target hardware address of the ARP
+// packet.
+func (c *Client) WriteTo(p *Packet, addr net.HardwareAddr) error {
+	pb, err := p.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	f := &ethernet.Frame{
+		Destination: p.TargetHardwareAddr,
+		Source:      p.SenderHardwareAddr,
+		EtherType:   ethernet.EtherTypeARP,
+		Payload:     pb,
+	}
+
+	fb, err := f.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	_, err = c.p.WriteTo(fb, &raw.Addr{HardwareAddr: addr})
+	return err
+}
+
+// Respond constructs and sends a response to an ARP request. On the
+// ARP layer, it will be addressed to the sender address of the
+// packet. On the ethernet layer, it will be sent to the actual remote
+// address from which the request was received.
+//
+// For more fine-grained control, use WriteTo to write a custom
+// response.
+func (c *Client) Respond(req *Packet, hwAddr net.HardwareAddr, ip net.IP) error {
```

Comment:
Alternatively it could also be SendResponse or SendReply (same question as for Request/SendRequest)